src/utils/ucc_coll_utils.c

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#include "ucc_coll_utils.h"
+#include "components/base/ucc_base_iface.h"
+#include "core/ucc_team.h"
+#define STR(x) #x
+#define STR_TYPE_CHECK(_str, _p, _prefix)       \
+    do {                                        \
+        if ((0 == strcasecmp(STR(_p), _str))) { \
+            return _prefix##_p;                 \
+        }                                       \
+    } while (0)
+
+#define STR_COLL_TYPE_CHECK(_str, _p) STR_TYPE_CHECK(_str, _p, UCC_COLL_TYPE_)
+
+ucc_coll_type_t ucc_coll_type_from_str(const char *str)
+{
+    STR_COLL_TYPE_CHECK(str, ALLGATHER);
+    STR_COLL_TYPE_CHECK(str, ALLGATHERV);
+    STR_COLL_TYPE_CHECK(str, ALLREDUCE);
+    STR_COLL_TYPE_CHECK(str, ALLTOALL);
+    STR_COLL_TYPE_CHECK(str, ALLTOALLV);
+    STR_COLL_TYPE_CHECK(str, BARRIER);
+    STR_COLL_TYPE_CHECK(str, BCAST);
+    STR_COLL_TYPE_CHECK(str, FANIN);
+    STR_COLL_TYPE_CHECK(str, FANOUT);
+    STR_COLL_TYPE_CHECK(str, GATHER);
+    STR_COLL_TYPE_CHECK(str, GATHERV);
+    STR_COLL_TYPE_CHECK(str, REDUCE);
+    STR_COLL_TYPE_CHECK(str, REDUCE_SCATTER);
+    STR_COLL_TYPE_CHECK(str, REDUCE_SCATTERV);
+    STR_COLL_TYPE_CHECK(str, SCATTER);
+    STR_COLL_TYPE_CHECK(str, SCATTERV);
+    return UCC_COLL_TYPE_LAST;
+}
+
+#define STR_MEM_TYPE_CHECK(_str, _p) STR_TYPE_CHECK(_str, _p,UCC_MEMORY_TYPE_)
+
+
+ucc_memory_type_t ucc_mem_type_from_str(const char *str)
+{
+    STR_MEM_TYPE_CHECK(str, HOST);
+    STR_MEM_TYPE_CHECK(str, CUDA);
+    STR_MEM_TYPE_CHECK(str, CUDA_MANAGED);
+    STR_MEM_TYPE_CHECK(str, ROCM);
+    STR_MEM_TYPE_CHECK(str, ROCM_MANAGED);
+    return UCC_MEMORY_TYPE_LAST;
+}
+
+static inline int
+ucc_coll_args_is_mem_symmetric(const ucc_base_coll_args_t *bargs)
```

Comment:
Naming: In general symmetric memory has a certain meaning, (at least coming from SHMEM world), it means that the address is same across all process. As we introduce PGAS collectives this might be more confusing. So, we should either 1) clearly say what symmetric memory here means. I’m not sure, where to capture though. 2) we could use other names ( ucc_coll_args_is_mem_type_homogenous  or ucc_coll_args_is_mem_type_same?).

Reply:
@manjugv agree with you. how about i call it "mem_type_mixed" ?